turbopack/crates/turbo-tasks-backend/src/kv_backing_storage.rs

```diff
@@ -646,14 +740,104 @@ fn serialize(task: TaskId, data: &Vec<CachedDataItem>) -> Result<SmallVec<[u8; 1
     })
 }
 
-fn deserialize_with_good_error<'de, T: Deserialize<'de>>(data: &'de [u8]) -> Result<T> {
-    match interning_serde::from_slice(&POT_CONFIG, data)? {
+fn deserialize_with_good_error<'de, D: KeyValueDatabase, T: Deserialize<'de>>(
+    database: &D,
+    tx: &D::ReadTransaction<'_>,
+    data: &'de [u8],
+) -> Result<T> {
+    let (global_ids, data) = LocalIdToGlobalId::read_from_slice(data)?;
+    let map = restore_strings(database, tx, &global_ids)?;
+
+    match interning_serde::from_slice::<T>(&POT_CONFIG, data, &map) {
         Ok(value) => Ok(value),
         Err(error) => serde_path_to_error::deserialize::<'_, _, T>(
             &mut pot_de_symbol_list().deserializer_for_slice(data)?,
         )
         .map_err(anyhow::Error::from)
-        .and(Err(error.into()))
+        .and(Err(error))
         .context("Deserialization failed"),
     }
 }
+
+/// Store the strings in the database and return the global ids.
+fn save_strings_serial<'a>(
+    batch: &mut impl SerialWriteBatch<'a>,
+    strings: &RcStrToLocalId,
+) -> Result<LocalIdToGlobalId> {
+    let mut global_ids = Vec::new();
+    for s in strings.iter() {
+        let (global_id, is_new) = get_string_id(batch, s)?;
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+/// Store the strings in the database and return the global ids.
+fn save_strings_concurrent<'a>(
+    batch: &impl ConcurrentWriteBatch<'a>,
+    strings: &RcStrToLocalId,
+) -> Result<LocalIdToGlobalId> {
+    let mut global_ids = Vec::new();
+    for s in strings.iter() {
+        let (global_id, is_new) = get_string_id(batch, s)?;
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+
+static STRING_INTERN_ID: OnceLock<AtomicU32> = OnceLock::new();
+
+/// Returns `(global_id, is_new)`
+fn get_string_id<'a>(batch: &impl BaseWriteBatch<'a>, s: &RcStr) -> Result<(u32, bool)> {
```

Comment:
This is currently broken and will write the same strings multiple times to the DB.

Reading from DB will not return values from the current write batch. And even if this would it would be a big race condition.

You need some synchronization map for already queried and written strings.

`DashMap<RcStr, u32>` would be fine. This could cache lookups and writes.